zipserver/archive.go

```diff
@@ -133,14 +136,22 @@ type UploadFileResult struct {
 	Size  uint64
 }
 
-func uploadWorker(a *Archiver, limits *ExtractLimits, tasks <-chan UploadFileTask, results chan<- UploadFileResult, cancel chan struct{}, done chan struct{}) {
+func uploadWorker(
+	ctx context.Context,
+	a *Archiver,
+	tasks <-chan UploadFileTask,
+	results chan<- UploadFileResult,
+	done chan struct{},
+) {
 	defer func() { done <- struct{}{} }()
 
 	for task := range tasks {
```

Comment:
Does it make sense to switch on the `ctx.Done()` here so it doesn't start a new job that would immediately be aborted if things have already been canceled?

Reply:
The current approach checks `ctx.Done()` before even producing a new task. Because the `tasks` channel is not buffered (production blocks on consumer readiness), we never produce tasks faster than the workers can complete them. Effectively, a worker won't start anything new once context is canceled.

So basically the problem should be solved on the producing end already, instead of the consuming end you've highlighted. Does that match your understanding?